### ISSUE TITLE
Extract "decorator utils" to separate addon.

### ISSUE THREAD

**Issue by rwjblue** [ember-decorators/ember-decorators#145]
I'm unsure what the name should be, but we should extract some of our building block utilities into a separate addon so that we can share the workload and allow other addons to easily create their own decorators.

The use cases that I am specifically thinking of are:

* @kellyselden's `ember-macro-helpers` (to provide the auto-get `@computed` version once it is deprecated/removed from here #146).
* @dfreeman's `ember-css-modules` based on https://github.com/rwjblue/ember-decorators/issues/133#issuecomment-316711907
* `ember-data` should provide its own decorators for `@attr` / `@hasMany` / etc.

**Comment by rwjblue** [comment]
One question I have is: should we move this repo to an org so that both ember-decorators and the new utils addon live in the same org?

**Comment by rwjblue** [comment]
The use cases that I was specifically thinking of for this issue are:

* @kellyselden's `ember-macro-helpers` (to provide the auto-get `@computed` version once it is deprecated/removed from here #146).
* @dfreeman's `ember-css-modules` based on https://github.com/rwjblue/ember-decorators/issues/133#issuecomment-316711907
* `ember-data` should provide its own decorators for `@attr` / `@hasMany` / etc.

**Comment by pzuraq** [comment]
I think `ember-decorator-macros` would work, that's essentially what they are. I like the idea of it being moved to an org, though it might be overkill to have an org just for the decorators. Would it be possible to use an existing one?

**Comment by rwjblue** [comment]
Not sure.  We have to be somewhat careful to not infer these addons are "official" while decorators are still in Stage 2, and moving the addon(s) to `emberjs` or `ember-cli` "feels" a bit to official to me...

**Comment by pzuraq** [comment]
yeah, fair enough. I'm down to create a new org, or we could move them to `html-next` possibly

**Comment by buschtoens** [comment]
I'm about to implement the aforementioned `ember-css-modules` decorators and would like to use these utils. I'm happy to do the manual labor of extracting these into their own addon, if you can't find the time. :smile: 

I suggest the name `@ember-decorators/utils` or do you have something better in mind?

**Comment by pzuraq** [comment]
Yep, that would be great. We actually already have the repo setup and published an earlier version: https://github.com/ember-decorators/utils

I was planning on using it to share metadata for `@ember-decorators/argument` but it turned out we didn't need to. Feel free to get rid of the -debug folder and stripping, etc.

Also, be aware that the utils will likely need to be rethought and reworked once the stage 2/3 transforms are finalized. That's why I've been putting it off myself, but I if you want to start now we can always refactor once they're finalized.

Also, do you want to make it an `ember-decorators` repo? `ember-decorators/css-modules` or something like that?

**Comment by buschtoens** [comment]
btw, should we also extract the babel configurator for the host app? If so, should I add it to the `utils` repo or create an extra repo for that?

https://github.com/ember-decorators/ember-decorators/blob/68362ab0ae3c7f2b7351ec31eac0f23a5bc38077/index.js#L52-L89

If you, as a user, install various decorator addons, you're most likely also installing `ember-decorators`, but there's no guarantee.

> I was planning on using it to share metadata for @ember-decorators/argument but it turned out we didn't need to. Feel free to get rid of the -debug folder and stripping, etc.

Any preferences regarding directory structure? Should I just dump all of [`utils`](https://github.com/ember-decorators/ember-decorators/tree/68362ab0ae3c7f2b7351ec31eac0f23a5bc38077/addon/utils) into the `addon` directory or should I also nest in `addon/utils`? I guess all decorators should be re-exported by `index.js`?

> Also, be aware that the utils will likely need to be rethought and reworked once the stage 2/3 transforms are finalized. That's why I've been putting it off myself, but I if you want to start now we can always refactor once they're finalized.

I guess that's the price we have to pay for living in the future. :stuck_out_tongue: 

> Also, do you want to make it an `ember-decorators` repo? `ember-decorators/css-modules` or something like that?

Good idea actually. This way I don't impose my preference for decorators onto all `ember-css-modules` users.

**Comment by pzuraq** [comment]
> btw, should we also extract the babel configurator for the host app? If so, should I add it to the utils repo or create an extra repo for that?

I think an extra repo would be good for that, `@ember-decorators/babel-transforms` or something like that. The tentative plan for v2 was to split out all of the decorators into individual packages to normalize the way they are imported and allow users to pick and choose the decorators they want, so `ember-decorators` itself would become a shell addon. If users didn't want all of the other addons, they would need the transforms addon as well.

> Any preferences regarding directory structure? Should I just dump all of utils into the addon directory or should I also nest in addon/utils? I guess all decorators should be re-exported by index.js?

I think dump them in `addon` directly, that way they can be imported from `@ember-decorators/utils`

Created both the repos and gave you write access, feel free to rename them if you want! Let me know when they're looking good and we can publish them 😄

**Comment by buschtoens** [comment]
> Created both the repos and gave you write access, feel free to rename them if you want! Let me know when they're looking good and we can publish them

Awesome, thank you! Did you intend to name the repo [`css-transforms`](https://github.com/ember-decorators/css-transforms) or was that by accident? :smile: 

Regarding `@ember-decorators/babel-transforms`: it this basically supposed to be a straight copy of [`index.js`](https://github.com/ember-decorators/ember-decorators/blob/68362ab0ae3c7f2b7351ec31eac0f23a5bc38077/index.js) and then marked as a `dependency` of all other `@ember-decorators/*`packages?

**Comment by pzuraq** [comment]
Yes, I’m afraid I’m not the best at names 😅 figured we could name it after the original library it wraps, but like I said feel free to change it, totally open here.

So the one tricky thing with babel transforms: currently they apply themselves directly to their parent, but if `ember-decorators` will include them we need a way to tell them to add themselves to the parent, and the parent of the parent as well

**Comment by buschtoens** [comment]
> Yes, I’m afraid I’m not the best at names  figured we could name it after the original library it wraps, but like I said feel free to change it, totally open here.

That's why I wanted to name it `css-modules` as well, but you created a repo called `css-transforms`. :stuck_out_tongue: 
Unfortunately I can't change the repo name. :see_no_evil: 


> So the one tricky thing with babel transforms: currently they apply themselves directly to their parent, but if ember-decorators will include them we need a way to tell them to add themselves to the parent, and the parent of the parent as well

Oh yeah. Didn't catch that. I guess we can do something like [`ember-font-awesome`](https://github.com/martndemus/ember-font-awesome/blob/38f24bc8b9c4b30c2922e40b35fa19d431b57ab0/index.js#L57-L62):

```js
  included(app, parentAddon) {
    // ...

    // Quick fix for add-on nesting
    // https://github.com/aexmachina/ember-cli-sass/blob/v5.3.0/index.js#L73-L75
    // see: https://github.com/ember-cli/ember-cli/issues/3718
    while (typeof app.import !== 'function' && (app.app || app.parent)) {
      app = app.app || app.parent;
    }

    // ...
  }
```

**Comment by pzuraq** [comment]
Ah sorry! Will change it, that was late last night lol

So font-awesome’s solution won’t work either, because we do actually only want to apply the transform to the parent. Every addon and the root app get their own independent version of ember-cli-babel, so if we apply it just to the root app it will _only_ be applied there, and everything else will break.

I think we could add an ember-cli option, `applyTransformsToParent`, that tells the addon to do this. Alternatively we could apply the transforms to the whole tree, but I think that would be aggressive

**Comment by dfreeman** [comment]
So given a dependency tree like:
```
my-cool-app
  -> some-addon
    -> @ember-decorators/argument
      -> @ember-decorators/babel-transforms
```

You're suggesting that if `@ember-decorators/argument` set `applyTransformsToParent: true`, then `@ember-decorators/babel-transforms` would turn them on for `some-addon`?

**Comment by pzuraq** [comment]
Yes, exactly. It could also nest, so `some-addon` could also specify the option and have it apply to the app

**Comment by dfreeman** [comment]
Technically I don't think there's anything that would prevent implementing that, but it does seem to run a little counter to the normal expectations for how build-time addons work. In general I'd have some slight concern about some n-times-nested addon changing my app's build configuration in a way I can't opt out of.

What happens if there are a bunch of instances of `@ember-decorators/babel-transforms` in my dependency tree, and they themselves have different resolved versions of the transform packages, all trying to activate for the root app? Which one wins?

**Comment by pzuraq** [comment]
That’a fair point, I wasn’t really expecting this feature to be used outside of `ember-decorators` but would more than likely happen if it’s public API 😜

The plugins do currently dedupe themselves, but that doesn’t solve opt-out. Maybe the functions can be exposed in the addon, and ember-decorators can import them and use them directly?

**Comment by pzuraq** [comment]
I personally really dislike default blueprints that just add peer dependencies, it’s something that’s always felt broken to me. But I suppose this would be a valid use case, and it would allow us to deprecate and remove the transforms as they become more standard

**Comment by dfreeman** [comment]
I honestly agree; I think most of the time just declaring a (non-`dev`) dependency is a much cleaner solution, and particularly addons from the earlier days of the ecosystem are guilty of abusing blueprints for that end.

This specific scenario, where you want to suggest use of an optional package that impacts the host rather than the addon itself, though, is the place I see the "blueprint for peer dependencies" approach as being warranted. My 2 cents ¯\\\_(ツ)_/¯

**Comment by buschtoens** [comment]
While stepping through the `_getParentOptions()` method, I noticed that it probably has a bug or didn't quite work as we think it would:

https://github.com/ember-decorators/ember-decorators/blob/37db49c687a44e456f9eddb5d855ed2298d75b67/index.js#L37-L50

`isEmberCLIProject` is a method, not a property. So the `if` statement ist always gonna execute the `else` branch.

Also, `included` is only ever called, if the `@ember-decorators/babel-transforms` addon is a direct child of the app / addon that is running the build pipeline. This means that in these scenarios `included` is calles:

```
example-addon ($ ember serve)
  -> @ember-decorators/babel-transforms
example-app ($ ember serve)
  -> @ember-decorators/babel-transforms
```

But not in these:

```
example-parent-addon ($ ember serve)
  -> example-addon
    -> @ember-decorators/babel-transforms
example-parent-app ($ ember serve)
  -> example-parent-addon
    -> example-addon
      -> @ember-decorators/babel-transforms
```

Also, `this.parent.options` never appears to be defined, while `this.app.options` always is.

But if I understand your comments and the ember-cli object graph correctly—properties don't seem to be documented at all—this doesn't really matter anymore anyway, since we now *always* want to read the config from the direct parent.

In conclusion we should be able to just go

```js
let parentOptions = this.app && this.app.options || {};
```

instead of

https://github.com/ember-decorators/ember-decorators/blob/37db49c687a44e456f9eddb5d855ed2298d75b67/index.js#L55

Or am I missing something?

**Comment by buschtoens** [comment]
Oh wait, I probably messed something. Didn't have my first coffee yet. :sweat_smile:

**Comment by pzuraq** [comment]
So, the `_getParentOptions` logic was due to a misconception I had based on the API docs. The API docs state that `included` only receives an EmberApp/EmberAddon, so I believed that it was only called from the root app and that addons were somehow flattened. This is not the case: https://github.com/ember-cli/ember-cli/blob/master/lib/models/addon.js#L655

So the `included` function signature should be `included(parent: EmberApp | EmberAddon | Addon)`, and parent is either the parent addon or app. As for the options, they may not be defined in the first place, which is why the `_getParentOptions` function creates and assigns them if they do not: https://github.com/ember-decorators/ember-decorators/blob/37db49c687a44e456f9eddb5d855ed2298d75b67/index.js#L44

So we should be able to just do:

```js
parentOptions = parent.options = parent.options || {};
```

And no worries, I too only see every other line of code before my morning joe 😝

**Comment by dfreeman** [comment]
It sounds like y'all have this under control, but a couple hard-learned lessons I've had to internalize working on this kind of stuff 😄 

- `this.parent` and the `parent` argument passed to included are two different things!
   - as @pzuraq called out, `included()` receives either an `EmberApp` instance or an `Addon` instance (or an `EmberAddon`, which is a subclass of `EmberApp`)
   - `this.parent` on an addon will either be an `Addon` or a `Project`, but never an `EmberApp`, so `this.parent.options` will actually miss any application options that were defined

- > Also, included is only ever called, if the @ember-decorators/babel-transforms addon is a direct child of the app / addon that is running the build pipeline.

  An addon's `included()` _should_ be called once per thing-that-depends-on-that-addon. In cases where you don't see it called for a nested addon, that usually means its parent isn't calling `super` correctly in its own `included()` implementation.

**Comment by buschtoens** [comment]
>> Also, included is only ever called, if the @ember-decorators/babel-transforms addon is a direct child of the app / addon that is running the build pipeline.

> An addon's included() should be called once per thing-that-depends-on-that-addon. In cases where you don't see it called for a nested addon, that usually means its parent isn't calling super correctly in its own included() implementation.

That was a blooper of mine. Instead of generating `example-addon` as an add-on, I generated it as an app. 

> It sounds like y'all have this under control

My manual tests seem to confirm this so far. However, I would *love* to add automatic tests for that (https://github.com/ember-decorators/babel-transforms/pull/3#issuecomment-345161791).

**Comment by pzuraq** [comment]
We've pretty much completed pulling out the utils at this point, and with v2 I'm pretty sure they've been factored down into some really common, reusable helper functions (at least until stage 2 drops 😛). Going to close this for now.
